programs/address-map/src/instruction.rs

```diff
@@ -0,0 +1,177 @@
+use {
+    crate::id,
+    serde::{Deserialize, Serialize},
+    solana_sdk::{
+        clock::Epoch,
+        instruction::{AccountMeta, Instruction},
+        pubkey::Pubkey,
+        system_program,
+    },
+};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum AddressMapInstruction {
+    /// Initialize an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Uninitialized address map account
+    ///   1. `[SIGNER, WRITE]` Account that will fund the new address map.
+    ///   2. `[]` System program for CPI.
+    InitializeAccount {
+        bump_seed: u8,
+        num_entries: u8,
+        authority: Pubkey,
+    },
+
+    /// Set a new authority for an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to set authority for
+    ///   1. `[SIGNER]` Current authority
+    SetAuthority { new_authority: Option<Pubkey> },
+
+    /// Insert entries into an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to insert entries into
+    ///   1. `[SIGNER]` Current authority
+    InsertEntries { offset: u8, entries: Vec<Pubkey> },
+
+    /// Activate an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to activate
+    ///   1. `[SIGNER]` Current authority
+    Activate,
+
+    /// Deactivate an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to deactivate
+    ///   1. `[SIGNER]` Current authority
+    Deactivate,
+
+    /// Close an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to close
+    ///   1. `[WRITE]` Recipient of closed account lamports
+    ///   2. `[SIGNER]` Current authority, if required.
+    CloseAccount,
+}
+
+/// Derive an address map address from a wallet address and the current epoch.
+pub fn derive_address_map_address(payer_address: &Pubkey, current_epoch: Epoch) -> (Pubkey, u8) {
+    Pubkey::find_program_address(
+        &[payer_address.as_ref(), &current_epoch.to_le_bytes()],
+        &id(),
+    )
+}
+
+/// Returns an instruction to initialize a map account as well as the address of
+/// the map account that will be created when the instruction is processed.
+pub fn initialize_account(
+    payer_address: Pubkey,
+    authority_address: Pubkey,
+    current_epoch: Epoch,
+    num_entries: u8,
+) -> (Instruction, Pubkey) {
+    let (map_address, bump_seed) = derive_address_map_address(&payer_address, current_epoch);
+    let instruction = Instruction::new_with_bincode(
+        id(),
+        &AddressMapInstruction::InitializeAccount {
+            bump_seed,
+            num_entries,
+            authority: authority_address,
```

Comment:
take it or leave it: Since we've had potential hijackings in the past, we may want to have a signature from the authority as well.  This way, bad integrations that just look for any account map for an authority can't get duped.

For example, if it's a defi application, the attacker sets up trap address map accounts for a bunch of addresses, using their own token accounts as the receivers, and just hopes that a bad integration uses the trap address map.

Reply:
Great call, I modified the create instruction here (https://github.com/solana-labs/solana/pull/21616) to use the authority in the PDA and enforce that it's a signer. It's also no longer possible to set the authority address to another address.